### ISSUE TITLE
Implement a tidy-like tool.

### ISSUE THREAD

**Issue by jyasskin** [tabatkins/bikeshed#662]
This should wrap lines, arrange elements sensibly, and indent `<pre>` blocks.

One convention I'm using in Web Bluetooth to reduce diffs is to put each sentence on its own line. I'm not certain this is worth maintaining, but the tidy tool could break after each `.` to help.

**Comment by tabatkins** [comment]
Semantic line-breaking is *absolutely* worth maintaining; I use it heavily.

**Comment by marcoscaceres** [comment]
It's the only reading I don't use BS :( I find it hard to keep code hand-organized - and it makes it hard for others to contribute in a clean way (the diffs end up a mess and element nesting becomes a nightmare). 

It would be awesome if it could somehow be made tidy compatible.

**Comment by tabatkins** [comment]
I don't quite understand - what exactly is the reason you don't use Bikeshed?

**Comment by marcoscaceres** [comment]
I can't tidy :(

**Comment by jyasskin** [comment]
Tidy itself is close and might be a good starting point. Its downsides are that:

* It adds a
```
<!DOCTYPE html>
<html>
  <head>
    <title></title>
  </head>
  <body>
```
  block at the top.

* It adds `=""` after any non-standard attribute.
* It wraps at a fixed column rather than incorporating any sort of semantic wrapping.
* It wraps `!Custom Keys` if they contain markup.

**Comment by marcoscaceres** [comment]
Ok, almost got this working in tidy 5.2... there is an option called "[show-body-only](http://api.html-tidy.org/tidy/quickref_5.2.0.html#show-body-only)" that treats input as if it's already wrapped in a `body` tag... so, that prevents the doctype, and other unnecessary garbage from being added. 

I've put in a feature request to HTML5 tidy to [prevent quoting of attributes](https://github.com/htacg/tidy-html5/issues/439) (or maybe just proprietary ones). I don't know if, for instance, having the "idl" attribute quoted in `<pre><a idl="">Thing</a></pre>` makes a difference to BS (probably not), but it would be nice for tidy to leave those alone too.

**Comment by marcoscaceres** [comment]
> * It wraps !Custom Keys if they contain markup.

@tabatkins, Maybe BS could be a bit smarted with these? allow them to be multiline? terminate on "^!", or "^.+:", or </pre>?

**Comment by marcoscaceres** [comment]
> * It wraps at a fixed column rather than incorporating any sort of semantic wrapping.

I like the idea of semantic wrapping (specially for markdown)... but I find it doesn't work in practice (for me) - specially in HTML. I like to rewrite and then move things around in paragraphs, to other sections, inside other tags that nest. I guess I'm just old, but having HTML structure/indenting helps me (I'm quite blind and a bit dyslexic) - so it all the tags, line breaks, an indenting actually helps me better see the structure of document as I'm writing it. 

Note that the "wrap:" option in tidy could be disabled or set to a large value to better support semantic wrapping... like setting it to 150 or larger.

**Comment by tabatkins** [comment]
> I don't know if, for instance, having the "idl" attribute quoted makes a difference to BS

It doesn't. I work off of a DOM for everything I possibly can, so I can't tell the difference between a boolean attribute and one with an empty string value.

> @tabatkins, Maybe BS could be a bit smarted with these? allow them to be multiline? terminate on "^!", or "^.+:", or `</pre>`?

I don't understand what you're asking.

> I guess I'm just old, but having HTML structure/indenting helps me (I'm quite blind and a bit dyslexic) - so it all the tags, line breaks, an indenting actually helps me better see the structure of document as I'm writing it.

I don't understand what any of this has to do with semantic line-wrapping.  It has nothing to do with tags or indenting.

**Comment by marcoscaceres** [comment]
> I don't understand what you're asking.

When I run tidy on this line:

```
!Participate: <a href="https://github.com/w3c/permissions">We are on Github.</a>
```

I get:
```
!Participate: <a href=
"https://github.com/w3c/permissions">We are on Github.</a>
```

Does that matter? Can BS handle that? 

> I don't understand what any of this has to do with semantic line-wrapping. It has nothing to do with tags or indenting.

Maybe you can educate me, because we clearly have different understanding of what SLW is (and yes, I've read the SLW document multiple times and I seem to be interpreting it differently to you in practice). 

Concretely, how would you say we deal with changing a `<p>` in the [Permissions spec](https://github.com/w3c/permissions/blob/master/index.bs), but retain our indenting relative to other tags and without content exceeding 80 cols?

**Comment by jyasskin** [comment]
http://rhodesmill.org/brandon/2012/one-sentence-per-line/ (not my source, but it captures the idea) just says to break lines at sentence boundaries and at logical points within a sentence, instead of reflowing to 80 columns. Changing the indentation level does produce diffs, but `-w` successfully ignores them, which it doesn't do if you reflow.

**Comment by tabatkins** [comment]
Yeah, it's the "80 columns" that's messing you up.  If you're using semantic line-breaking, you *don't* apply another line-breaking strategy. (But most of the time the lines do end up being short enough to fit in.

If you're stuck under an 80 columns restriction, then I can't help you. Those restrictions are stupid and harmful for plain-text files, and they produce *terrible* diffs from trivial changes.  Try, if possible, to abandon that and just embrace natural line-breaking.  If you check out the source of some CSS specs written by fantasai or me, you'll see we break at ends of sentences, at commas and similar punctuation, and at many joiners (like "and" or "or").

(And @jyasskin's link is indeed my original source.)

**Comment by jyasskin** [comment]
It's easy enough to hold to an 80 column limit *plus* semantic line breaks. Web Bluetooth does this (at 100 columns). It means that when you go over the column limit, you look for an extra place to break. As long as you also break at sentence and other semantic boundaries, the length limit doesn't reflow the rest of the paragraph.

**Comment by tabatkins** [comment]
> Does that matter? Can BS handle that?

In general, no. Metadata blocks allow for continuation lines, but it's as if you started that line with the same tag again.  So in other words,

```
!Participate: <a href=
 "https://github.com/w3c/permissions">We are on Github.</a>
```

(note that the continuation lines need to start with some whitespace)

is equal to

```
!Participate: <a href=
!Participate: "https://github.com/w3c/permissions">We are on Github.</a>
```

So that's not what you want.  This is just going to be a problem for anything that naively treats Bikeshed files as HTML.

**Comment by domfarolino** [comment]
Thanks @jyasskin for pointing out this issue. It's probably worth mentioning that I started writing https://github.com/domfarolino/specfmt, which is a simple rust command line tool that tries to format web specs (mostly just be rewrapping). It has two main modes:
 - Format an entire spec file
 - Format lines that appear in the git diff of your current branch

It's mostly a rewrapper based on https://github.com/domenic/rewrapper with some improvements. It's not a real "parser" or anything, and it's probably pushing the limits of a non-real-parser stab at formatting. Still might be worth integrating into the bikeshed toolchain somehow...

**Comment by tabatkins** [comment]
As I noted earlier, I'm strongly opposed to anything that auto-wraps text at a particular column width. It's actively harmful to diffs. WHATWG people have the exact opposite policy, and that's fine for them, but I won't ever be adding any tooling to Bikeshed that supports such a wrapping strategy.

**Comment by domfarolino** [comment]
We could always add a semantic line wrapping mode, and make that the default for Bikeshed invocations of it. I'm certainly open to that.
